File: 0045-jump-game-ii/0045-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        memo = {}
        def dp(index):
            if index >= len(nums)-1:
                return 0
            an = float("inf")
            for i in range(1, nums[index]+1):
                if index + i not in memo:
                    memo[index+i] = 1+dp(index+i)
                an = min(an,memo[index+i])
            return an
        return dp(0)

File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        vector<int> dp;
        for(int i=0; i < nums.size(); i++){
            dp.push_back(214748364);
        }
        dp[dp.size()-1] = 0;
            
        for(int i=nums.size()-2; i > -1 ; i--){
            for(int j=1; j <= nums[i] ; j++){
                if (i+j >= nums.size()){
                    break;
                }
                dp[i] = min(dp[i], 1+dp[i+j]);
            }
        }
        return dp[0];
    }
};